amqprs/src/api/channel/basic.rs

```diff
@@ -549,7 +605,10 @@ impl Channel {
         let (consumer_tx, mut consumer_rx): (
             mpsc::Sender<ConsumerMessage>,
             mpsc::Receiver<ConsumerMessage>,
-        ) = mpsc::channel(CONSUMER_MESSAGE_BUFFER_SIZE);
+        ) = mpsc::channel(std::cmp::max(
+            CONSUMER_MESSAGE_BUFFER_SIZE,
+            self.prefetch_count().into(),
+        ));
```

Comment:
I understand your intention, but we cannot easily solve such config issue easily. Note that, buffer happens in different level:  Connection -> Channels -> Consumers, we do not know how many channels user will create in advance on a connection, and we do not know how many consumers user will create in advance on a channel.  

- we cannot calculate the max buffer size for connection and channel. 
- Connection and channel buffer is configured before consumer, we can't really change that based on prefetch count.
- Meanwhile, the tokio bounded channel will not fail just because the number of send is higher than buffer size, it just block and wait for available permit, I said it would overflow is when user hold the receiver but never use it, then will cause  overflow.
- Above all, the internal channel buffer is different from server's prefetch count, we may have more messages buffered in client than prefetch count due to auto-ack mode, rely on prefetch count to config the buffer is not reliable in this way.

Reply:
From what I understand reading the docs, the rmq client only needs to know what the `prefetch_count` per consumer is - it doesn't matter what the max prefetch count for a given channel is, since consumer qos takes priority. I also understand that the mspc channel is created per Consumer, so this matches with notion of pretech, i.e. the rmq server will never send more  than `prefetch_count` unacked messages.

> we cannot calculate the max buffer size for connection and channel.

I'm not sure I understand what you mean, is this linked to what I'm currently doing ?

> Connection and channel buffer is configured before consumer, we can't really change that based on prefetch count.

It doesn't really matter in some sense, the library controls consumption for these buffers and then the dispatcher will send the message to the consumer's rx. Once it becomes the library's user's responsibility to consume messages, we have no control anymore and cannot guarantee that it will be done in a timely manner such that it'll avoid overflows. That's why I'm proposing to size the mspc channel based on server constraints.

> user hold the receiver but never use it, then will cause overflow.

Wouldn't that also happen if a user takes a lot of time (>10mins) to ack/nack each message ?

> we may have more messages buffered in client than prefetch count due to auto-ack mode, rely on prefetch count to config the buffer is not reliable in this way.

That's why I set the channel's size to a max between prefetch_count and the buffer size constant

Reply:
There is also a setting in rmq for max channel per connection : https://www.rabbitmq.com/channels.html#channel-max

As I did with the call to `basic_qos`, you can set it once the user has called it and use a default value close enough to the server's default.
When a user starts tweaking these settings, it usually means that they know what they're doing and will need to tune the client appropriately.

To me this is unrelated to the matter of prefetch and sizing the consumer's buffer though.

Reply:
> From what I understand reading the docs, the rmq client only needs to know what the `prefetch_count` per consumer is - it doesn't matter what the max prefetch count for a given channel is, since consumer qos takes priority. I also understand that the mspc channel is created per Consumer, so this matches with notion of pretech, i.e. the rmq server will never send more than `prefetch_count` unacked messages.

consumer is created on channel, there can be multiple consumers created on single channel.
Also I meant the messaing buffering is in different level Connection -> Channels -> Consumers **within the library**, connection buffering all  messages on connectin level, then channel buffering all messages on channel level.
If start consumer as `no_ack`, which means auto-ack, it is possible that the client already ack all the messages but not yet processing all the messages.

Reply:
> Wouldn't that also happen if a user takes a lot of time (>10mins) to ack/nack each message ?
> 
> >

Not necessary, because when messages exceed the bounded channel's capacity, tokio will sleep at `send().await` (a.k.a blocking), it will not necessary overflow.
Meanwhile, even setting the buffer size according to prefetch count, still possible the arriving messages exceed the bounded channel capacity, similar reason as I mentioned above regarding async scheduling,  and RMQ's auto-ack mode

Reply:
> There is also a setting in rmq for max channel per connection : https://www.rabbitmq.com/channels.html#channel-max
> 
> As I did with the call to `basic_qos`, you can set it once the user has called it and use a default value close enough to the server's default. When a user starts tweaking these settings, it usually means that they know what they're doing and will need to tune the client appropriately.
> 
> To me this is unrelated to the matter of prefetch and sizing the consumer's buffer though.

What I meant above all is that just rely on prefetch count to configure consumer buffer size cannot easily solve the buffering mistmatch. Let's take an example

Example 1 - Assume we create a channel (default buffer size 256), then start a consumer with config prefetch size = 256, then we start a another consumer with config prefetch size = 256. What should we do?

Example 2 - Assume we create a channel (default buffer size 256), then set global prefetch count = 1024, how can we change the channel buffer after channel created?

Example 3 - Let's say to magically know in advance channel max = 1024, pretech size for channel is 1000, max consumers = 10, consumer prefetch count = 100,  then do we really want to allocate connection's buffer = 1024*1000*10*100? it is a huge waste.

So, you might already see that even configure the consumer buffer size according to consumer prefetch count, messages can be buffered on upstream level buffering, e.g. connection leve or channel level

Reply:
> handling of all the messages have been auto acked.

This problem can arise whatever the value of the buffer size is; the core issue being that the channel is bounded (even if it were unbounded, at some point we could have an OOM error). To me the library can only do so much and the user must be aware of what he is doing.

There may be a misunderstanding. I'm talking about the buffer size of the consumer mspc channel. This is the one that we should be careful about given consumption is not owned by the library, especially in the case of `basic_consume_rx` where you return the `Receiver`.
The connection & the channel's mspc buffer can be configured to any size, it will always be read (as long as there are no internal errors). I may be wrong, but to me if all consumers have the maximum unacked messages set with `basic_qos` the server will not send more messages, even if the channel allows for more.

The problem I see here is that given we have a bounded channel, what would be the optimal way to size the buffer correctly. The C# client uses unbounded channels, what would you think about switching from `tokio::sync::mspc::channel` to `futures::sync::mspc::unbounded`?

Reply:
> The connection & the channel's mspc buffer can be configured to any size, it will always be read (as long as there are no internal errors). I may be wrong, but to me if all consumers have the maximum unacked messages set with `basic_qos` the server will not send more messages, even if at the channel level qos settings allow for more.

Please see the Example 2, which channel already backpresure first, not the consumer,  

> The problem I see here is that given we have a bounded channel, what would be the optimal way to size the buffer correctly. The C# client uses unbounded channels, what would you think about switching from `tokio::sync::mspc::channel` to `futures::sync::mspc::unbounded` just for the consumer?

I would still prefer bounded channel to avoid memory usage grow wildly, this ensure the library is more reliable, while providing APIs to allow user to tune the buffering according to their application. 

What are the advantages & drawbacks of setting consumer buffer size simply according to prefetch count compared to providing APIs for user to tune all the buffering size?